File: autodom/proxy.py
"""Domcom's Decodo datacenter route, with its residential route as fallback."""

import os
from base64 import b64encode
from dataclasses import dataclass, field
from urllib.parse import urlsplit


@dataclass(frozen=True, slots=True)
class ProxyRoute:
    tier: str
    url: str
    authorization: str = field(repr=False)
    port_start: int = 0
    port_count: int = 0

    def url_for(self, page: int) -> str:
        if not self.port_count:
            return self.url
        port = self.port_start + (page - 1) % self.port_count
        return self.url.rsplit(":", 1)[0] + f":{port}"


def _route(tier: str, endpoint: str, username: str, password: str) -> ProxyRoute:
    if not username or not password:
        raise ValueError(f"Configure the {tier} SMARTPROXY username and password")
    try:
        parsed = urlsplit(endpoint if "://" in endpoint else "http://" + endpoint)
        if (
            parsed.scheme not in {"http", "https"}
            or not parsed.hostname
            or parsed.port is None
            or parsed.username is not None
            or parsed.path not in {"", "/"}
            or parsed.query
            or parsed.fragment
        ):
            raise ValueError
        if ":" in username:
            raise ValueError
        authorization = "Basic " + b64encode(f"{username}:{password}".encode("latin1")).decode(
            "ascii"
        )
        port_start = int(os.environ.get(f"SMARTPROXY_{tier.upper()}_PORT_START", "0"))
        port_count = int(os.environ.get(f"SMARTPROXY_{tier.upper()}_PORT_COUNT", "0"))
        if port_count < 0 or (
            port_count and not 1 <= port_start <= port_start + port_count - 1 <= 65535
        ):
            raise ValueError
    except (ValueError, UnicodeError):
        raise ValueError(f"Invalid {tier} SMARTPROXY endpoint or authentication format") from None
    return ProxyRoute(
        tier, parsed._replace(path="").geturl(), authorization, port_start, port_count
    )


def load_proxy_routes() -> tuple[ProxyRoute, ProxyRoute]:
    """Require both configured tiers. Never fall back to direct internet access."""
    username = os.environ.get("SMARTPROXY_USERNAME", "").strip()
    password = os.environ.get("SMARTPROXY_PASSWORD", "").strip()
    datacenter = _route(
        "datacenter",
        os.environ.get("SMARTPROXY_DATACENTER_ENDPOINT")
        or os.environ.get("SMARTPROXY_ENDPOINT")
        or "dc.decodo.com:10000",
        username,
        password,
    )
    residential = _route(
        "residential",
        os.environ.get("SMARTPROXY_RESIDENTIAL_ENDPOINT") or "gate.decodo.com:7000",
        os.environ.get("SMARTPROXY_RESIDENTIAL_USERNAME", "").strip(),
        os.environ.get("SMARTPROXY_RESIDENTIAL_PASSWORD", "").strip(),
    )
    return datacenter, residential

File: packages/core/src/proxy.ts
import { inspect } from "node:util";

type Environment = Readonly<Record<string, string | undefined>>;

export class ProxyRoute {
  constructor(
    readonly tier: string,
    readonly url: string,
    readonly authorization: string,
    readonly port_start = 0,
    readonly port_count = 0,
  ) {}

  urlFor(page: number): string {
    if (!this.port_count) return this.url;
    const port =
      this.port_start + ((((page - 1) % this.port_count) + this.port_count) % this.port_count);
    return `${this.url.slice(0, this.url.lastIndexOf(":"))}:${port}`;
  }

  [inspect.custom](): string {
    return `ProxyRoute { tier: ${JSON.stringify(this.tier)}, url: ${JSON.stringify(this.url)}, port_start: ${this.port_start}, port_count: ${this.port_count} }`;
  }
}

function route(
  tier: string,
  endpoint: string,
  username: string,
  password: string,
  env: Environment,
): ProxyRoute {
  if (!username || !password)
    throw new Error(`Configure the ${tier} SMARTPROXY username and password`);
  try {
    const value = endpoint.includes("://") ? endpoint : `http://${endpoint}`;
    const parsed = new URL(value);
    // WHATWG URL removes explicit default ports, so recover and require the written port.
    const authority = /^[a-z]+:\/\/([^/?#]*)/iu.exec(value)?.[1] ?? "";
    const portText = /:(\d+)$/u.exec(authority)?.[1];
    const port = Number(portText);
    if (
      !["http:", "https:"].includes(parsed.protocol) ||
      !parsed.hostname ||
      !portText ||
      port < 1 ||
      port > 65535 ||
      authority.includes("@") ||
      !["", "/"].includes(parsed.pathname) ||
      parsed.search ||
      parsed.hash ||
      username.includes(":") ||
      /[\u0100-\u{10ffff}]/u.test(username + password)
    )
      throw new Error();
    const startText = (env[`SMARTPROXY_${tier.toUpperCase()}_PORT_START`] ?? "0").trim();
    const countText = (env[`SMARTPROXY_${tier.toUpperCase()}_PORT_COUNT`] ?? "0").trim();
    if (!/^[+-]?\d+$/u.test(startText) || !/^[+-]?\d+$/u.test(countText)) throw new Error();
    const port_start = Number(startText);
    const port_count = Number(countText);
    if (
      !Number.isSafeInteger(port_start) ||
      !Number.isSafeInteger(port_count) ||
      port_count < 0 ||
      (port_count > 0 && (port_start < 1 || port_start + port_count - 1 > 65535))
    )
      throw new Error();
    const authorization = `Basic ${Buffer.from(`${username}:${password}`, "latin1").toString("base64")}`;
    return new ProxyRoute(
      tier,
      `${parsed.protocol}//${parsed.hostname}:${port}`,
      authorization,
      port_start,
      port_count,
    );
  } catch {
    throw new Error(`Invalid ${tier} SMARTPROXY endpoint or authentication format`);
  }
}

export function loadProxyRoutes(env: Environment = process.env): readonly ProxyRoute[] {
  const datacenter = route(
    "datacenter",
    env.SMARTPROXY_DATACENTER_ENDPOINT || env.SMARTPROXY_ENDPOINT || "dc.decodo.com:10000",
    (env.SMARTPROXY_USERNAME ?? "").trim(),
    (env.SMARTPROXY_PASSWORD ?? "").trim(),
    env,
  );
  const residential = route(
    "residential",
    env.SMARTPROXY_RESIDENTIAL_ENDPOINT || "gate.decodo.com:7000",
    (env.SMARTPROXY_RESIDENTIAL_USERNAME ?? "").trim(),
    (env.SMARTPROXY_RESIDENTIAL_PASSWORD ?? "").trim(),
    env,
  );
  return [datacenter, residential];
}
